### PR TITLE
feat: Unit tests for exactly once. Covers scenario with message specific errors

### PR DESCRIPTION
**Pull request by Rishabh-V**
Hi Jon. This is not-urgent and can be reviewed next week once you are back. I have modified the mocks for exactly once handling and added couple of more tests for message specific errors. Thanks

### CODE REVIEW

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs [hunk 1]
```diff
@@ -324,11 +324,7 @@ public override async Task AcknowledgeAsync(SubscriptionName subscription, IEnum
                 // Non-gRPC exceptions will propagate to the caller.
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:324]: Please modify line the comments above, as they're no longer accurate.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs [hunk 2]
```diff
@@ -362,6 +354,54 @@ public override async Task ModifyAckDeadlineAsync(SubscriptionName subscription,
...
+            /// <list type="bullet">
+            /// <item> Checks if subscription under test is exactly once or not.</item>
+            /// <item> If it's not exactly once, it will always throw the specified exception for the specified number of times as <c>_numberOfAckModifyAckDeadlineFailures</c>.</item>
+            /// <item> If it's exactly once, it first checks if the entire request has failed permanently based on <c>StatusCode</c>. If yes, it will set the flag to throw the specified exception.</item>
+            /// <item> If the entire request has not failed permanently, it will check the error metadata to see if it contains permanent error for the given ackId. If yes, it will set the flag to throw the specified exception.</item>
+            /// <item> If the flag to throw the specified exception is set and the specified number of exceptions are not thrown already, the specified exception is thrown, else exception is not thrown.</item>
+            /// </list>
+            /// </summary>
+            /// <param name="ackIds">The list of acknowledgement ids being processed.</param>
+            private void MayBeThrowException(IEnumerable<string> ackIds)
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:369]: Nit: MayBe => Maybe (as maybe is one word)

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs [hunk 3]
```diff
@@ -362,6 +354,54 @@ public override async Task ModifyAckDeadlineAsync(SubscriptionName subscription,
...
+            {
+                // For non-exactly once subscriptions, we always throw exception.
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:371]: Rather than keeping a local variable for this, I would target the *one* situation where you don't want to do anything (exactly-once delivery, an RpcException that is retriable and none of the ack IDs has a permanent error), and return at that point. I think that would be clearer.

In particular, it would make it more obvious that in this situation we *don't* modify `_numberOfAckModifyAckDeadlineFailures` - I'm assuming that was deliberate. If you *did* intend to call `CompareExchange` (and maybe increment `_numberOfAckModifyAckDeadlineFailures`) unconditionally, I'd do that at the start of the method.

Basically at the moment it's not clear to me whether the expected behavior is "modify state and maybe throw an exception" or "maybe ignore completely, but otherwise modify state and throw an exception".

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs [hunk 4]
```diff
@@ -416,6 +456,31 @@ public void Dispose()
...
+                errorInfo = new Rpc.ErrorInfo { Domain = "pubsub.googleapis.com", Reason = "broken" };
+                // Message 2 is temporary failure, 3 is permanent failure.
+                errorInfo.Metadata.Add("2", "TRANSIENT_FAILURE_UNORDERED_ACK_ID");
+                errorInfo.Metadata.Add("3", "PERMANENT_FAILURE_INVALID_ACK_ID");
+            }
+
+            var status = new Google.Rpc.Status { Details = { Any.Pack(errorInfo) } };
+            var metadata = new Metadata { { "grpc-status-details-bin", status.ToByteArray() } };
+            return new RpcException(new Status(StatusCode.OK, ""), metadata);
+        }
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:483]: Nit: add blank line before the next test

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs [hunk 5]
```diff
@@ -416,6 +456,31 @@ public void Dispose()
...
+        /// <param name="errorInfo">The <see cref="Rpc.ErrorInfo"/> instance with metadata populated with message id and corresponding failure.
+        /// Can be null, in which case the default <see cref="Rpc.ErrorInfo"/> initialized inside in this method is used,
+        /// which marks message id 2 as temporary failure and message id 3 as permanent failure.</param>
+        /// <returns>The <see cref="RpcException"/> with temporary and permanent errors.</returns>
+        /// <remarks>
+        /// This test method is useful to test exactly once subscription only. 
+        /// </remarks>
+        private static RpcException GetExactlyOnceMixedException(Rpc.ErrorInfo errorInfo = null)
+        {
+            if (errorInfo == null)
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:471]: I'd do all of this as:

```csharp
errorInfo ??= new Rpc.ErrorInfo
{
    Domain = "pubsub.googleapis.com",
    Reason = "broken",
    Metadata =
    {
        { "2",   "TRANSIENT_FAILURE_UNORDERED_ACK_ID" },
        { "3",   "PERMANENT_FAILURE_INVALID_ACK_ID" }
    }
};
```

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs [hunk 6]
```diff
@@ -1231,5 +1296,75 @@ public void ExactlyOnce_Extends_PermanentFault()
...
+            var ackModifyAckDeadlineAction =
+                ackOrNack ? AckModifyAckDeadlineAction.BadAck(exception, numberOfFailures: 1)
+                : AckModifyAckDeadlineAction.BadModifyAckDeadline(exception, numberOfFailures: 1);
+
+            using var fake = Fake.Create(new[] { msgs }, useMsgAsId: true, ackModifyAckDeadlineAction: ackModifyAckDeadlineAction, isExactlyOnce: true);
+
+            fake.Scheduler.Run(async () =>
+            {
+                var doneTask = fake.Subscriber.StartAsync(async (msg, ct) =>
+                await Task.FromResult(ackOrNack ? SubscriberClient.Reply.Ack : SubscriberClient.Reply.Nack));
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1327]: Indent this line to show it's part of the StartAsync call.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs [hunk 7]
```diff
@@ -1231,5 +1296,75 @@ public void ExactlyOnce_Extends_PermanentFault()
...
+                await fake.TaskHelper.ConfigureAwait(fake.Scheduler.Delay(TimeSpan.FromSeconds(100), CancellationToken.None));
+                Exception ex = await fake.TaskHelper.ConfigureAwaitHideErrors(() => fake.Subscriber.StopAsync(CancellationToken.None));
+                // For acks or nacks, exception should be thrown for only permanent failure.
+                Assert.True(ex is not null);
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1331]: I'd just use Assert.NotNull for this. (Or just the IsType check below, which I believe already checks for non-nullity).

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs [hunk 8]
```diff
@@ -1231,5 +1296,75 @@ public void ExactlyOnce_Extends_PermanentFault()
...
+                var doneTask = fake.Subscriber.StartAsync(async (msg, ct) =>
+                {
+                    // Add delay greater than ackDeadline to simulate the extends.
+                    await fake.TaskHelper.ConfigureAwait(fake.Scheduler.Delay(TimeSpan.FromSeconds(70), ct));
+                    return await Task.FromResult(SubscriberClient.Reply.Ack);
+                });
+                await fake.TaskHelper.ConfigureAwait(fake.Scheduler.Delay(TimeSpan.FromSeconds(100), CancellationToken.None));
+                Exception ex = await fake.TaskHelper.ConfigureAwaitHideErrors(() => fake.Subscriber.StopAsync(CancellationToken.None));
+                // For extends, exception shouldn't be thrown.
+                Assert.True(ex is null);
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1366]: Assert.Null for this.

**Review by jskeet — COMMENTED** [review]
I confess I haven't looked in detail at the PubSub unit tests for some time - I don't fully understand the MaybeThrowException aspect. (It feels like a lot of this is effectively trying to mirror the behavior elsewhere, which rings some alarm bells but is probably unavoidable.) The general aim seems okay though.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 1 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:325]: Updated the comment.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 2 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs]: Done.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 3 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:374]: I have added the retryable scenario without permanent error as suggested by you. I have also updated the comments to make the intention of the method clear. Please let me know if further changes are needed and I am happy to do so.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 5 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs]: Done.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 4 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:495]: Done.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 6 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs]: Done.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 7 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs]: Done.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 8 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs]: Done.

**Review by Rishabh-V — COMMENTED** [review]
Hi Jon! I have addressed the review comments in the new commit. I have made updates to `MaybeThrowException` to make it clearer but not entirely sure if I am able to incorporate everything that you want. Please take a look and I can make further changes based on your feedback.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 3 above)
Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:374]: It still feels to me like there's a lot of code (30 lines) which is basically "should we immediately return" - but it's not written that way. (And in fact at the moment in one case you *are* returning, and in another you're just setting `throwException` to false, when they're equivalent - and that's confusing.)

**Review by jskeet — COMMENTED** [review]
Side note: the tests at the moment seem to be testing whether an exception is being thrown. Are we also testing that the relevant IDs are being retried? I can't immediately see that, but it's entirely possible that's just because the testing here is quite intricate.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs (hunk 3 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:374]: Sure Jon. Understood. I will try to refactor it in next commit and we can take it further from there.

**Review by Rishabh-V — COMMENTED** [review]
Right Jon. We do test the retry part. It's just that I couldn't come up with an Assert that makes it explicit. Will try to add an Assert statement in next commit along with the refactoring.

**Comment by Rishabh-V** [comment]
Closing this for #8973. Will request review for it next week.
